cli/QuickstepCli.cpp

```diff
@@ -135,9 +141,22 @@ int main(int argc, char* argv[]) {
   gflags::ParseCommandLineFlags(&argc, &argv, true);
   google::InitGoogleLogging(argv[0]);
 
-  if (quickstep::FLAGS_num_workers > 0) {
-    printf("Starting Quickstep with %d worker thread(s)\n",
-           quickstep::FLAGS_num_workers);
+  // Detect the hardware concurrency level. Note this call will return 0
+  // if it fails (which it may on some machines/environments).
+  const unsigned int num_hw_threads = std::thread::hardware_concurrency();
+
+  // Use the command-line value if that was supplied, else use the value
+  // that we computed above, provided it did return a valid value.
+  // TODO(jmp): May need to change this at some point to keep one thread
+  //            available for the OS if the hardware concurrency level is high.
```

Comment:
No, it would actually be the number of SMT threads. So, on my laptop with
hyperthreading (4 cores, 8 hyperthreads), it returns 8.

Hardware SMT threads appear as physical CPU cores to the operating system
(sort of... Linux actually does detect which of the virtual "cores" are SMT
siblings and has some scheduling optimizations based on that... for
instance it assumes that the cost of migrating a thread to its sibling is
negligible, but that's not visible to userspace programs).

We should also be clear about what is meant by "maximum concurrency". Linux
and other modern OSes have no problem with scheduling thousands of threads
on a CPU core (although if the threads in question are high CPU use or
contend for other resources, you can expect the system to chug).
std::thread::hardware_concurrency() just gives us a hint about the actual
number of parallel CPU cores (or with SMT, thread contexts) *in hardware*.
Since our workers are mostly either CPU or memory bound, it makes good
sense to use that number of workers. Old performance experiments showed
that when the workload was mostly CPU bound with few cache misses, using
the number of CPU cores vs. doubling to use hyperthreads made almost no
difference in performance. On the other hand, when memory access was an
issue and cache misses were higher, we got a decent boost (10-30% IIRC)
from using the hyperthreads (mainly because while one of the virtual cores
in a pair was stalling during a cache miss, the other one could
almost-instantly take over the physical core and get some useful work done).

TLDR: std::thread::hardware_concurrency() is just a "hint" about
concurrency. It's certainly safe to go over what it says, but the hint is
actually quite good for our purposes, whether or not hyperthreading is
available.

On Thu, Jan 21, 2016 at 11:07 AM Zuyu ZHANG <notifications@github.com>
wrote:

> In cli/QuickstepCli.cpp
> <https://github.com/pivotalsoftware/quickstep/pull/7#discussion_r50446799>
> :
>
> > @@ -135,9 +141,22 @@ int main(int argc, char* argv[]) {
> >    gflags::ParseCommandLineFlags(&argc, &argv, true);
> >    google::InitGoogleLogging(argv[0]);
> >
> > -  if (quickstep::FLAGS_num_workers > 0) {
> > -    printf("Starting Quickstep with %d worker thread(s)\n",
> > -           quickstep::FLAGS_num_workers);
> > +  // Detect the hardware concurrency level. Note this call will return 0
> > +  // if it fails (which it may on some machines/environments).
> > +  const unsigned int num_hw_threads = std::thread::hardware_concurrency();
> > +
> > +  // Use the command-line value if that was supplied, else use the value
> > +  // that we computed above, provided it did return a valid value.
> > +  // TODO(jmp): May need to change this at some point to keep one thread
> > +  //            available for the OS if the hardware concurrency level is high.
>
> Cool. So if a CPU supports simultaneous multithreading (SMT), the max
> number of concurrent threads would be 2x of the number of CPU cores
> returned by std::thread::hardware_concurrency?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pivotalsoftware/quickstep/pull/7/files#r50446799>.
>